### PR TITLE
feat: Sheet plugin (with virtual grid)

### PR DESCRIPTION
**Pull request by richburdon**
Virtual grid experiment using `react-window` and `hyperformula`.

## Next steps

- [ ] Row/Column numbers
- [ ] Resize Columns
- [ ] Enable selection range via keyboard (shift-left/right/up/down, etc.)
- [x] Automerge/Codemirror editor

https://github.com/user-attachments/assets/2c137145-5854-4aa6-8dbd-b746c54ff3a5

### CODE REVIEW

File: packages/apps/plugins/plugin-grid/src/components/Matrix/context.tsx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+export const useMatrixEvent = () => {
+  const { event } = useMatrixContext();
+  return event;
+};
+
+export type CellValue = string | number | undefined;
+
+export const MatrixContextProvider = ({ children, data }: PropsWithChildren<{ data?: CellValue[][] }>) => {
+  const [event] = useState(new Event<CellEvent>());
```
Comment (ZaymonFC) [inline packages/apps/plugins/plugin-grid/src/components/Matrix/context.tsx:66]:
```suggestion
  const event = useMemo(() => new Event<CellEvent>());
```

**Comment by bvaughn** [comment]
I'm seeing failed Storybook assertions whenever I try to edit a cell after updating to commit 403e6ac.

File: packages/apps/plugins/plugin-grid/src/components/Matrix/context.tsx (hunk 1 above)
Comment (richburdon) [reply packages/apps/plugins/plugin-grid/src/components/Matrix/context.tsx:77]: isn't memo not guaranteed to get called again (thus losing any bound events?)

**Comment by richburdon** [comment]
> I'm seeing failed Storybook assertions whenever I try to edit a cell after updating to commit [403e6ac](https://github.com/dxos/dxos/commit/403e6ac495915ba68a1fb840f58852918c997a43).

I pushed a fix for this. 
Also have replaced the raw input with a CodeMirror instance with basic formula styling (e.g., for ranges).

**Comment by bvaughn** [comment]
> > I'm seeing failed Storybook assertions whenever I try to edit a cell after updating to commit [403e6ac](https://github.com/dxos/dxos/commit/403e6ac495915ba68a1fb840f58852918c997a43).
> 
> I pushed a fix for this. Also have replaced the raw input with a CodeMirror instance with basic formula styling (e.g., for ranges).

Nice.

Not sure if you saw my email, but I put together a proof of concept this afternoon built on top of [Lezer](https://lezer.codemirror.net/) and [Lexical](https://lexical.dev/) that adds a light-weight spreadsheet editor (with auto-complete and syntax highlighting)
 
Kapture 2024-08-02 at 17 45 07

**Comment by richburdon** [comment]
> > > I'm seeing failed Storybook assertions whenever I try to edit a cell after updating to commit [403e6ac](https://github.com/dxos/dxos/commit/403e6ac495915ba68a1fb840f58852918c997a43).
> > 
> > 
> > I pushed a fix for this. Also have replaced the raw input with a CodeMirror instance with basic formula styling (e.g., for ranges).
> 
> Nice.
> 
> Not sure if you saw my email, but I put together a proof of concept this afternoon built on top of [Lezer](https://lezer.codemirror.net/) and [Lexical](https://lexical.dev/) that adds a light-weight spreadsheet editor (with auto-complete and syntax highlighting)
> 
> Kapture 2024-08-02 at 17 45 07 [ Kapture 2024-08-02 at 17 45 07 ](https://private-user-images.githubusercontent.com/29597/354774833-cd15a6ee-7aee-4861-ba26-b47199a43132.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2NTc2NDYsIm5iZiI6MTcyMjY1NzM0NiwicGF0aCI6Ii8yOTU5Ny8zNTQ3NzQ4MzMtY2QxNWE2ZWUtN2FlZS00ODYxLWJhMjYtYjQ3MTk5YTQzMTMyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDAzNTU0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5YzFiZDFlN2UxMjZhMTQyZGRkZTdjNzZmZGE0OWVmNmVlMmE2MjllZTQyNTQ0ZTc5ZjlkNDlkOWI1NmNkZDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FUBNe4-EKtfTe4zAszp4FnDg0_IuisrdPNYHaA5_Rn8) [ ](https://private-user-images.githubusercontent.com/29597/354774833-cd15a6ee-7aee-4861-ba26-b47199a43132.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjI2NTc2NDYsIm5iZiI6MTcyMjY1NzM0NiwicGF0aCI6Ii8yOTU5Ny8zNTQ3NzQ4MzMtY2QxNWE2ZWUtN2FlZS00ODYxLWJhMjYtYjQ3MTk5YTQzMTMyLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODAzVDAzNTU0NlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTk5YzFiZDFlN2UxMjZhMTQyZGRkZTdjNzZmZGE0OWVmNmVlMmE2MjllZTQyNTQ0ZTc5ZjlkNDlkOWI1NmNkZDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FUBNe4-EKtfTe4zAszp4FnDg0_IuisrdPNYHaA5_Rn8)

oh wow, that's great. do you have a PR -- are you able to rebase on this current branch -- i moved things around quite a bit, so hopefully the merge isn't annoying.

**Comment by bvaughn** [comment]
> do you have a PR -- are you able to rebase on this current branch? (i moved things around quite a bit, so hopefully the merge isn't annoying).

I have been rebasing occasionally. Lots of conflicts in `Cell.tsx` and the PNPM lockfile (unsurprisingly) but no big deal.

**Comment by richburdon** [comment]
> > do you have a PR -- are you able to rebase on this current branch? (i moved things around quite a bit, so hopefully the merge isn't annoying).
> 
> I have been rebasing occasionally. Lots of conflicts in `Cell.tsx` and the PNPM lockfile (unsurprisingly) but no big deal.

I've merged the PR and I'll step away for the moment to work on the backend Functions -- so things will be more stable now.

**Pull request merged**